### PR TITLE
local preprocess

### PR DESCRIPTION
**Pull request by yohm**
Add a new function "local preprocess" to Simulator and Analyzer.
If a user specifies a command to this field, the command is executed at the host where OACIS is running before submitting the job to a remote host.
With this new functionality, users can

- conduct a preprocess which is not available on a remote host
- copy results of another job.
    - In other words, users can define a job which depends on another job although it requires a fairly complex coding using APIs.

## Usage

Fill in "local pre process script" field when defining Simulator or Analyzer
image

The command is executed before submitting at the directory of the Run or Analysis. The output files of the local pre-process will be copied to the working directory of the remote host.

### CODE REVIEW

File: app/services/remote_job_handler.rb [hunk 1]
```diff
@@ -83,44 +123,60 @@ def prepare_input_json(job)
...
+      copy_files_to_work_dir_via_ssh(job, org_dest_list)
     end
   end
 
-  def prepare_input_files_via_copy(job)
-    remote_path = RemoteFilePath.input_files_dir_path(@host,job)
+  def filelist_of_local_preprocess(job)
+    org_dest_list = []
+    Dir.chdir(job.dir) {
+      org_dest_list = Dir.glob("**/*").map {|f| [ Pathname.new(f).expand_path, f ] }
```
Comment (takeshi-uchitane) [inline app/services/remote_job_handler.rb]: Analysis has a field named files_to_copy to select some specified files. Here, whole files and directories are going to be copied. Is it a desired specification?

File: app/services/remote_job_handler.rb [hunk 2]
```diff
@@ -66,6 +69,43 @@ def set_submitted_to_if_necessary(job)
     end
   end
 
+  def execute_local_pre_process(job)
+    return unless job.executable.local_pre_process_script.present?
+    Dir.chdir( job.dir ) {
```
Comment (takeshi-uchitane) [inline app/services/remote_job_handler.rb]: Local pre process runs on public/Result_development/`executable.id`. It means that some files may be overwritten if remote jobs create files with the same name. Is it a desired specification?

Comment (yohm) [inline app/services/remote_job_handler.rb]: This is an intensional behavior. All the output files of local preprocess are updated by the output files of the simulator.

File: app/services/remote_job_handler.rb (hunk 1 above)
Comment (yohm) [inline app/services/remote_job_handler.rb]: This is an intensional behavior as well. All the output files of a local pre-process are copied to the remote work directory. If you do not want to copy a certain file, you may remove it in the local pre-process.

**Pull request merged**